### PR TITLE
Add .dot graph visualization of AstExecGraph

### PR DESCRIPTION
**Pull request by mortbopet**
This commit adds the creation of .dot files - used by GraphViz - to visualize how mtasks are statically scheduled across the set of specified threads.
We visualize each thread as a row, with nodes of a row being the mtasks scheduled for the given thread. The width of the mtask nodes are proportional to their cost. mtask dependencies are shown using an edge between the source and sink mtasks.

It could be argued that this information would be more easily displayed in a .vcd file. However, i find this rather restrictive in terms of not being able to annotate ie. mtask dependencies. Furthermore, keeping this to .dot files allows for embedding more details in the future, if the multithread scheduling implementation is elaborated.

### CODE REVIEW

**Comment by mortbopet** [comment]
Here's some example outputs:
t3
t24

[t_altera_lpm_24threads.pdf](https://github.com/verilator/verilator/files/5927832/t_altera_lpm_24threads.pdf)
[t_gen_alw_3threads.pdf](https://github.com/verilator/verilator/files/5927834/t_gen_alw_3threads.pdf)

**Review by wsnyder — COMMENTED** [review]
Nicely done, that's a nice layout for the data too.

Bonus points - walk the graph to determine critical path (starting at time zero any path with any arrows that don't go over any "gaps" in time) and color edges and boxes on it red.  On your larger example should make it much easier to visualize what's important.

File: src/V3AstNodes.cpp [hunk 1]
```diff
@@ -219,6 +219,96 @@ AstExecGraph::AstExecGraph(FileLine* fileline)
...
+}
+
+//! Variant of dumpDotFilePrefixed without --dump option check
+void AstExecGraph::dumpDotFilePrefixedAlways(const string& nameComment) const {
+    dumpDotFile(v3Global.debugFilename(nameComment) + ".dot");
+}
+
+void AstExecGraph::dumpDotFile(const string& filename) const {
+    if (!debug()) return;
+
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply src/V3AstNodes.cpp:234]: Thanks - When i started the implementation i briefly looked at https://github.com/verilator/verilator/blob/master/src/V3Partition.cpp#L2250 where this part must have stuck around.

**Comment by mortbopet** [comment]
Working on implementing critical path estimation for the AstExecGraph, i've realized that the above implementation does not correctly reflect how the mtasks are scheduled if the m_depGraphp is not a connected graph (which is most often the case).
I don't think there is any way around storing the estimated start- and end times within the mtasks themselves. This is both useful for estimating the critical path locally in AstExecGraph as well as properly visualizing the scheduling.

image

**Comment by mortbopet** [comment]
@wsnyder 
> Bonus points - walk the graph to determine critical path (starting at time zero any path with any arrows that don't go over any "gaps" in time) and color edges and boxes on it red. On your larger example should make it much easier to visualize what's important.

Regarding this, i am not sure that we can define a critical path in this way with the current scheduling algorithm. 
Take, for instance, this dependency graph:
image
mt102 is a long running computation, and a dependency of mt147. Due to how `PartPackMTask::completionTime` is used, this results in mt147 being scheduled on thread 1 some time after the completion of mt102 on thread 0.

image

To remedy this, i suspect we can detect the critical path if we, starting from the mtask scheduled to finish last (mt147 in the above), walk the graph by walking towards the dependency which is scheduled to finish _last_.

**Comment by wsnyder** [comment]
>i suspect we can detect the critical path if we, starting from the mtask scheduled to finish last

Yes, this seems reasonable.

We should probably update one of the test pl files to dump (--dump-tree) and compares the dot output with a golden reference. Perhaps copy t/t_threads_nondeterminism.pl to do this?

**Pull request closed without merging**